### DISCUSSION TITLE
API Endpoint to Download DAM Settlement Point Prices | NP4-190-CD

### DISCUSSION THREAD

**Discussion by suvo1004** [ercot/api-specs#48] [Q&A] [answered]
Hello Team, 

I am trying to download DAM Settlement Point Prices report using a Java Plugin. I was able to get the EndtPoint URL following the instruction provided https://developer.ercot.com/applications/pubapi/    -   [[https://api.ercot.com/api/public-reports/np4-190-cd/dam_stlmnt_pnt_prices](https://www.google.com/url?q=https://urldefense.com/v3/__https:/www.google.com/url?q%3Dhttps:**Aurldefense.com*v3*__https:*www.google.com*url*q*3Dhttps:**AAapi.ercot.com*api*public-reports*np4-190-cd*dam_stlmnt_pnt_prices*26source*3Dgmail-imap*26ust*3D1725291094000000*26usg*3DAOvVaw0CGxj1vJu349VOmab_7gjw__;Ly8vLy8v!!DR3VkBMYqM1H!e-Gn5fMdEMcR6O_Reo7Qa0_Kvo1EdIIhTRyaCQjaAEHF3NCojySkNKIM-NS8onATVZXDZIpYuMyOA7_BnjIBOISEU8ZY$%26source%3Dgmail-imap%26ust%3D1725292140000000%26usg%3DAOvVaw1X3_WP-_LB2eAlB9Yk-Qvr__;Ly8vLy8vPyUqKioqKiolJSUlJSU!!DR3VkBMYqM1H!aPd9yg_zP4x1r6zgw4fsO1IiHh89DKjV50rDlmbjq3z5mvaTGKC09Xet6KvpwsfEJO6Ai2B2-38M0QdmVmCismKb8Fyq$&source=gmail-imap&ust=1725304807000000&usg=AOvVaw1PyKh-5pwOK73YOH94O_tt)](Endpoint URL)

**_Response received:_**
`{
  "emilId": "NP4-190-CD",
  "name": "DAM Settlement Point Prices",
  "description": "The Settlement Point Prices for all Resource Nodes, Load Zones, and Trading Hubs from the Day-Ahead Market for the last thirty days on a daily basis.",
  "status": "Active",
  "reportTypeId": 12331,
  "audience": "Public",
  "generationFrequency": "Event - Per DAM Run",
  "securityClassification": "Public",
  "lastUpdated": "2021-02-07",
  "firstRun": "2010-11-29",
  "eceii": null,
  "channel": "Public, EWS",
  "userGuide": null,
  "postingType": "Report",
  "market": "Nodal",
  "extractSubscriber": null,
  "xsdName": "Current Day Reports XSD",
  "misPostingLocation": null,
  "certificateRole": null,
  "fileType": "zip, csv, xml",
  "ddlName": null,
  "misDisplayDuration": 31,
  "archiveDuration": 2555,
  "notificationType": null,
  "contentType": "DATA",
  "downloadLimit": 500,
  "lastPostDatetime": "2024-08-22T12:32:40",
  "protocolRules": {
    "NP4.5.3(2)(b)": "https://www.ercot.com/mp/data-products?protocolRules=NP4.5.3(2)(b)"
  },
  "artifacts": [
    {
      "reportTypeId": 12331,
      "displayName": "DAM Settlement Point Prices",
      "_links": {
        "endpoint": {
          "href": "https://api.ercot.com/api/public-reports/np4-190-cd/dam_stlmnt_pnt_prices"
        }
      }
    }
  ],
  "_links": {
    "self": {
      "href": "https://api.ercot.com/api/public-reports/np4-190-cd"
    },
    "parent": {
      "href": "https://api.ercot.com/api/public-reports"
    },
    "archive": {
      "href": "https://api.ercot.com/api/public-reports/archive/np4-190-cd"
    }
  }
}`

But when I am trying to download the file from the EndPoint using Java Input/Output Steam, I am getting the file structure as content instead of actual SPP data  

`{
  "report": {
    "reportName": "dam_stlmnt_pnt_prices",
    "reportDisplayName": "DAM Settlement Point Prices",
    "reportId": "12331",
    "reportEMIL": "NP4-190-CD",
    "downloadLimit": 2000000
  },
  "fields": [
    {
      "name": "deliveryDate",
      "label": "Delivery Date",
      "cardinality": 1,
      "dataType": "DATE",
      "searchable": true,
      "sortable": true,
      "hasRange": true
    },
    {
      "name": "hourEnding",
      "label": "Hour Ending",
      "cardinality": 2,
      "dataType": "VARCHAR",
      "searchable": true,
      "sortable": true,
      "hasRange": false
    },
    {
      "name": "settlementPoint",
      "label": "Settlement Point",
      "cardinality": 3,
      "dataType": "VARCHAR",
      "searchable": true,
      "sortable": true,
      "hasRange": false
    },
    {
      "name": "settlementPointPrice",
      "label": "Settlement Point Price",
      "cardinality": 4,
      "dataType": "DOUBLE",
      "searchable": true,
      "sortable": true,
      "hasRange": true
    },
    {
      "name": "DSTFlag",
      "label": "DST Flag",
      "cardinality": 5,
      "dataType": "BOOLEAN",
      "searchable": true,
      "sortable": true,
      "hasRange": false
    }
  ],
  "_links": {
    "self": {
      "href": "https://api.ercot.com/api/public-reports/np4-190-cd/dam_stlmnt_pnt_prices"
    },
    "parent": {
      "href": "https://api.ercot.com/api/public-reports/np4-190-cd"
    }
  }
}`

Can someone please help with how I can fix the issue, many thanks.

**Comment by mackermann-ercot** [comment, marked as answer]
Hello suvo1004, 

I have a few questions, if you don't mind. 

   1. What is the exact URL you are using?  (including parameters)
   2. Do you have a code snippet you can share so we can get a better idea of how you are attempting to download the file?
   3. Are you using the "download" parameter, or simply trying to read the HTTP via openStream() to obtain an InputStream?
   4.  Are you specifying the appropriate media type ("Accept:" header) for the data?

If you don't mind the file being buffered into memory before it is written to the filesystem, you can use simple Java calls (no libraries):

```
InputStream in = new URL(FILE_URL).openStream();
Files.copy(in, Paths.get(FILE_NAME), StandardCopyOption.REPLACE_EXISTING);
```

An example URL used with this code (for your example) could be:

```
https://api.ercot.com/api/public-reports/np4-190-cd/dam_stlmnt_pnt_prices?deliveryDateFrom=2024-08-27&download
```
In this example the data is being retrieved for all records from today at midnight, and in a ZIP file format.  The default data file within the archive is JSON, however you can change the format by adding the format value ("json" or "csv"):

```
... &download=csv
```

**Comment by suvo1004** [comment]
Hello  Mackermann, 

Thank you very much for your response. 
I was not using any Parameter while downloading the file which was the root cause. Once I update the URL to include the parameter suggested by you, it worked. Thanks again for your time and have a nice day ahead!

**Reply by mackermann-ercot** [reply]
That's excellent.  I'm glad we could help you solve the issue!

Regards,
mackermann-ercot
